presto-plan-checker-providers/src/main/java/com/facebook/presto/plancheckerproviders/nativechecker/NativePlanChecker.java

```diff
@@ -0,0 +1,159 @@
+import com.facebook.presto.spi.ConnectorId;
+import com.facebook.presto.spi.Node;
+import com.facebook.presto.spi.NodeManager;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.TableHandle;
+import com.facebook.presto.spi.WarningCollector;
+import com.facebook.presto.spi.plan.PlanChecker;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.PlanVisitor;
+import com.facebook.presto.spi.plan.SimplePlanFragment;
+import com.facebook.presto.spi.plan.TableScanNode;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.Set;
+
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static com.facebook.presto.spi.StandardErrorCode.NOT_FOUND;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Uses the native sidecar to check verify a plan can be run on a native worker.
+ */
+public final class NativePlanChecker
+        implements PlanChecker
+{
+    private static final Logger LOG = Logger.get(NativePlanChecker.class);
+    private static final MediaType JSON_CONTENT_TYPE = MediaType.parse("application/json; charset=utf-8");
+    public static final String PLAN_CONVERSION_ENDPOINT = "/v1/velox/plan";
+
+    private final NodeManager nodeManager;
+    private final JsonCodec<SimplePlanFragment> planFragmentJsonCodec;
+    private final OkHttpClient httpClient;
+
+    public NativePlanChecker(NodeManager nodeManager, JsonCodec<SimplePlanFragment> planFragmentJsonCodec)
+    {
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+        this.planFragmentJsonCodec = requireNonNull(planFragmentJsonCodec, "planFragmentJsonCodec is null");
+        this.httpClient = new OkHttpClient.Builder().build();
+    }
+
+    private URL getPlanValidateUrl()
+    {
+        Set<Node> nodes = nodeManager.getAllNodes();
+        Node sidecarNode = nodes.stream().filter(Node::isCoordinatorSidecar).findFirst()
+                .orElseThrow(() -> new PrestoException(NOT_FOUND, "could not find native sidecar node"));
+        try {
+            // endpoint to retrieve session properties from native worker
+            return new URL(sidecarNode.getHttpUri().toString() + PLAN_CONVERSION_ENDPOINT);
+        }
+        catch (IOException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "I/O Error opening native sidecar location", e);
+        }
+    }
+
+    @Override
+    public void validate(PlanNode planNode, WarningCollector warningCollector)
+    {
+        // NO-OP, only validating fragments
+    }
+
+    @Override
+    public void validateFragment(SimplePlanFragment planFragment, WarningCollector warningCollector)
+    {
+        if (!planFragment.getPartitioning().isCoordinatorOnly() && !isInternalSystemConnector(planFragment.getRoot())) {
```

Comment:
@rschlussel continuing the discussion from https://github.com/prestodb/presto/pull/23596#discussion_r1783411780

With these checks removed there are a few PlanFragments that fail in the sidecar conversion. I have a hive connector and when starting the CLI, I issues this query

```
SELECT table_name FROM information_schema.tables WHERE table_schema = 'tpch';
```
The native plan checker fails with:
```
Query 20241017_222152_00003_fr9ih failed: Error response from native plan checker with code: 500, response: { "status": "error", "message": "Protocol for connector $info_schema not registered")}
```

Some other queries that fail:
```
presto:tpch> select * from system.runtime.nodes;
Query 20241017_221219_00008_xj86v failed: Error response from native plan checker with code: 500, response: { "status": "error", "message": "(nodes vs. tasks) SystemConnector supports only tasks table")}

presto:tpch> SHOW SCHEMAS FROM system;
Query 20241017_220528_00003_xj86v failed: Error response from native plan checker with code: 500, response: { "status": "error", "message": "Protocol for connector $info_schema not registered")}
```

With the checks enabled, these are not sent to the sidecar, although `isCoordinatorOnly()` doesn't seem to be set on any of these so doesn't make a difference.

Reply:
why do these fail with the plan checker? What happens if you just run them on a prestissimo cluster without the native plan checker?  if it fails there too, then it's fine to fail with the plan checker, and  if it doesn't we should get to the bottom of what's different with the native plan checker and fix it properly.

Reply:
i think it would be good to check how the code works for filtering out these fragments from being sent to workers in the regular code path and see if we can reuse it.

I also think we should remove that check in InternalPlanNode.accept().  I don't see a good reason that a PlanVisitor shouldn't be able to operate on InternalPlanNodes.  That should resolve your issue here to be able to handle complex plans.

Reply:
> i think it would be good to check how the code works for filtering out these fragments from being sent to workers in the regular code path and see if we can reuse it.

I did step through it to the point where it tries to load the system connector and it is not registered in the sidecar and it can't continue the conversion at that point.

> I also think we should remove that check in InternalPlanNode.accept(). I don't see a good reason that a PlanVisitor shouldn't be able to operate on InternalPlanNodes. That should resolve your issue here to be able to handle complex plans in a robust way.

Sure, I can remove the InternalPlanNode check and retest.

Reply:
I think there needs to be some more discussion about the check in `InternalPlanNode`. Reading the original [commit](https://github.com/prestodb/presto/commit/7bd6ecac2f513f819c6ff88a3cffd79f2b92470b) message, it is intentionally there to prevent connector access to certain plan nodes from the SPI, I'm guessing there is a security concern. Also, if the check for an `InternalPlanVisitor` is removed, then there is no point to have `InternalPlanNode` or `InternalPlanVisitor` at all. @tdcmeehan do you have any thoughts on this?

Reply:
yeah, I've been thinking about this independent of this change, as we've been looking at adding other custom plan checkers. If we are going to have the ability to have custom plan checkers, we need to be able to iterate through the plan.  I don't think there's a real security concern here (it's basically a convenience method that we're missing. the connector still gets the full plan and can write a more functional style visitor if it wanted).  The connector would only have access to methods that are in the SPI (e.g. getSources()).

I also find that it's pretty arbitrary which PlanNodes are in the SPI and which are InternalPlanNodes.  It seems like PlanNodes get added to the SPI whenever it seems like it would be useful. I sort of wonder if we should just move all plan nodes to the SPI and get rid of the InternalPlanNode concept altogether. I spot checked a bunch of examples and all the fields of the plan nodes are already in the spi, so I'm not sure what we're getting out of having these two concepts.

Reply:
If I remember correctly, the distinction between InternalPlanNode vs. PlanNode was for connector plan optimizers: we would never want to send certain plan nodes to a connector plan optimizer, because it would most likely indicates a bug.
 
But I think now, we have essentially expanded the scope of the PlanNode to include everything, because we are serializing it and giving it to an external system, Velox.  So in my view, I think we can just move everything over to PlanNode.

I think it should be done in this PR, because as @rschlussel says, we definitely will serialize `InternalPlanNode`s, so the checker will need access to them.  We can add it as a new commit at the beginning.

Reply:
i created a separate PR for just this change https://github.com/prestodb/presto/pull/23876

Reply:
sgtm, I will rebase and update

Reply:
done, and looks like everything is working correctly